pkg/policy/distillery_test.go

```diff
@@ -0,0 +1,311 @@
+			// If there's no L4 policy, undecided becomes allow.
+			if len(*l4IngressPolicy) == 0 && !endpointSelected {
+				key := Key{uint32(id), 0, 0, dirIngress}
+				result[key] = MapStateEntry{l7RedirectNone_}
+				io.WriteString(d.log, "[distill] Endpoint not selected; allowing\n")
+			}
+			// Otherwise this will be handled in L4 resolution below.
+		case api.Denied:
+			io.WriteString(d.log, "[distill] L3 ingress denied\n")
+			deniedPeers[id] = lbls
+		}
+	}
+
+	// Handle L4 ingress from each identity in the cache to the endpoint.
+	io.WriteString(d.log, "[distill] Producing L4 filter keys\n")
+	for _, l4 := range *l4IngressPolicy {
+		io.WriteString(d.log, fmt.Sprintf("[distill] Processing L4Filter (l7: %+v)\n", l4.L7RulesPerEp))
+		for _, key := range l4.ToKeys(0, d.identityCache, deniedPeers) {
+			io.WriteString(d.log, fmt.Sprintf("[distill] L4 ingress allow %+v (parser=%s, redirect=%t)\n", key, l4.L7Parser, l4.IsRedirect()))
+			if l4.IsRedirect() {
+				result[key] = MapStateEntry{l7RedirectProxy}
+			} else {
+				result[key] = MapStateEntry{l7RedirectNone_}
+			}
+		}
+	}
+
+	// Handle L3-wildcard of L7 destinations
+	// Eg, when you have L4+L7 "allow /public on 80" with L3 "allow all from foo"
+	// Initially, three keys would be generated: L4+L7, L3+L4+L7, and L3.
+	// Here, we remove the L3+L4+L7 key if it overlaps with the L4+L7,
+	// but only if they have the same L7 redirect.
+	//
+	// For these the BPF policy order of attempting L3+L4 lookup, L4 lookup,
+	// then L3 lookup means that three keys are not strictly necessary; the
+	// correct behaviour can be encoded with two keys - an L4 key and L3 key.
+	nKeys := 0
+	l3l4keys := make([]Key, len(result))
+	for k := range result {
+		if k.Identity > 0 && k.DestPort > 0 {
+			l3l4keys[nKeys] = k
+			nKeys++
+		}
+	}
+	for i := 0; i < nKeys; i++ {
+		k := l3l4keys[i]
+		io.WriteString(d.log, fmt.Sprintf("[distill] Squashing L3-dependent L4 key %+v\n", k))
+		wildcardL3 := Key{DestPort: k.DestPort, Nexthdr: k.Nexthdr, TrafficDirection: k.TrafficDirection}
+		wildcardL4 := Key{Identity: k.Identity, TrafficDirection: k.TrafficDirection}
+		if _, ok := result[wildcardL4]; ok {
+			io.WriteString(d.log, fmt.Sprintf("[distill] -> Found L3 overlap %+v\n", wildcardL4))
+			if entry, ok := result[wildcardL3]; ok {
+				io.WriteString(d.log, fmt.Sprintf("[distill] -> Found L4 overlap %+v:%+v\n", wildcardL3, entry))
+				if entry.ProxyPort == result[k].ProxyPort {
+					io.WriteString(d.log, fmt.Sprintf("[distill] -> Removing L3-dependent L4 %+v\n", k))
+					delete(result, k)
+				}
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func Test_MergeRules(t *testing.T) {
+	identityCache := cache.IdentityCache{
+		identity.NumericIdentity(identityFoo): labelsFoo,
+	}
+
+	tests := []struct {
+		test   int
+		rules  api.Rules
+		result MapState
+	}{
+		//  Rule 0                   | Rule 1         | Rule 2         | Rule 3         | Rule 4         | Desired BPF map state
```

Comment:
nit: It might be helpful to put the google doc URL actually in the code here, rather than only in the commit message.